torch/distributed/_symmetric_memory/__init__.py

```diff
@@ -467,3 +472,224 @@ def restride_A_for_fused_matmul_reduce_scatter(
     return make_contiguous_for_perm(t, perm)
+
+
+class Work(_Work):
+    def __init__(self) -> None:
+        super().__init__()
+        self.event = torch.cuda.Event()
+        self.event.record()
+
+    def wait(self, timeout: timedelta = timedelta(seconds=0)) -> bool:
+        self.event.wait()
+        return True
+
+
+# NOTE [low-contention collectives]
+# When a collective is overlapped with abundant compute, it makes sense to
+# prioritize reducing the contention between the collective and the overlapped
+# compute, even at the cost of a slightly slower collective.
+#
+# Common collective implementations (e.g., NCCL without user buffer
+# registration) optimize for throughput with no ambient compute. However, such
+# implementations may not be optimal when they are overlapped with compute:
+# - These impls typically fuse the entire collective into a single kernel and
+# reserve SM resources based on the most demanding portion of the collective,
+# even when a large portion of the collective does not require this much
+# resource.
+# - These implementations typically fuse the entire collective into a single
+# kernel and reserve SM resources based on the most demanding portion of the
+# collective, even when a large portion of the collective does not require this
+# much resource.
+# - These implementations often use SM-based P2P copy as opposed to copy
+# engine-based P2P copy. Copy engine-based P2P copy may not have a significant
+# advantage when there's no ambient compute. However, it may significantly
+# improve overall resource utilization in the presence of ambient compute.
+#
+# When overlapped with intensive compute (e.g., persistent matmul kernels), the
+# SM-usage of a collective can lead to inefficient overlapping.
+#
+# Low-contention collectives achieve their goals with the following strategies:
+# - Use copy engine-based copy whenever possible.
+# - Break down portions of a collective with different resource requirements
+# into multiple kernels. This improves the overlapping efficiency at the cost
+# of additional launching overhead.
+@torch.library.impl(lib, "_low_contention_all_gather", "Meta")
+def _low_contention_all_gather_meta(
+    tensor: torch.Tensor,
+    group_name: str,
+) -> torch.Tensor:
+    group_size = c10d._get_group_size_by_name(group_name)
+    return tensor.new_empty(tensor.shape[0] * group_size, *tensor.shape[1:])
+
+
+@torch.library.impl(lib, "_low_contention_all_gather", "CUDA")
+def _low_contention_all_gather(
+    tensor: torch.Tensor,
+    group_name: str,
+) -> torch.Tensor:
+    """
+    Performs all-gather with symmetric memory in a low-contention fashion.
+
+    When `tensor` is already in symmetric memory:
+        - The collective is carried out without using SMs.
+        - No symmetric memory workspace is required.
+
+    When `tensor` is not in symmetric memory:
+        - An extra SM-based copy is performed to copy the input data into the
+          symmetric memory workspace.
```

Comment:
Sorry, I missed that these symm. mem-based collectives follow the funcol signature. I _think_ that should not be a problem since the stream synchronization should ensure correctness regardless.

The only thing to watch out for is that in our current implementation, the AG input is a view into the AG output.